### PR TITLE
Dockercompose for frontoffice

### PR DESCRIPTION
**Pull request by dguillen12**
## Description
refs #17 

Please provide enough information so that others can review your pull request:

Modify Docker-compose for frontoffice

### CODE REVIEW

File: nginx/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+events { worker_connections 1024; }
+
+http {
+    server {
+        listen 80;
+        root  /usr/share/nginx/html;
+        include /etc/nginx/mime.types;
+
+        location / {
+            try_files $uri /index.html;
```
Comment (eudago) [inline nginx/nginx.conf:12]: add proxy pass to node server with port 3000

**Review by eudago — CHANGES_REQUESTED** [review]
nginx shows "works" from index.html, it would be necessary to point to the nuxt server

**Review by d-asensio — CHANGES_REQUESTED** [review]
Thank you @dguillen12!

Taking into account that we will be deploying this service to a k8s cluster, I assume that this docker-compose is going to be used just for development purposes, so departing from this premise:

Do we really need Nginx? In development, this is not adding any value, at least in this case that we only have a single service.

I think that it would be better to simplify this, what do you think @jbagot ?

In addition, if this is meant for development, we should map the local directory to the image so we can actually change files while developing.

Regarding the production Docker image, we have to take into account that this is not going to be a static page, we want to make use of the Server Side Rendering capabilities of Nuxt, so we will need to run the app instead of generating it.

Maybe It'd be better to have a quick chat about the necessities we have for this service, I think that this will be faster 😄

**Comment by jbagot** [comment]
Hi, yesterday we did this together with Didac and Eric. We don't know how works nuxt and we found some articles, someones created an nginx like us with the statics inside the nginx, and others using the development server in port 3000.
With the comments of @eudago and @d-asensio I think we have to proxy to the server in port 3000. Then my question is, in production we will need the same? We will build and run nuxt with yarn and this will create the server and then with a reverse proxy we will have to point to this one? We thought that this server was a development purpose server and we should use static files inside nginx instead.
The other David's comment about using nginx in the compose it's true that is not needed, but we did it for 2 reasons, first to learn how to do it and second it obligated us to create another Dockerfile that will be the one used in k8s. Then I think it's not a bad practice to have this docker-compose to push the image to the registry, for example. For development purposes you have to use the `frontoffice` service and skip nginx. But we will create pipelines that will push the nginx image to the registry and we will use it in the k8s.

**Comment by d-asensio** [comment]
> With the comments of @eudago and @d-asensio I think we have to proxy to the server in port 3000. Then my question is, in production we will need the same? We will build and run nuxt with yarn and this will create the server and then with a reverse proxy we will have to point to this one? We thought that this server was a development purpose server and we should use static files inside nginx instead.

To clarify this I will describe the two possible workflows, one for development and the other for production:

## Development

- Install dependencies by running `yarn`
- Run the development server `yarn dev`

The development server helps us by hot-reloading the browser every time we make a change, so we don't have to re-run any command.

> To perform hot-reloading, Nuxt injects code into the javascript bundle that is served to the browser. Such code establishes a WebSocket connection (in the same port) with the server (that is watching for changes in the filesystem) and every time a change is notified the browser receives a message and downloads a chunk of code to update itself, you can see this in the following image:
>
>  image


## Production

- Install dependencies by running `yarn`
- Generate the optimized production build `yarn build`, this creates the `.nuxt` folder (that is the production artifact and auto-contains all the required dependencies). 
- Run the server, either by using `yarn start` or `nuxt start` in the same directory where the generated `.nuxt` folder is placed.

Essentially the only required thing to run the Nuxt application in production mode is the generated `.nuxt` artifact and the nuxt-cli, that can be installed globally in the system through `export NUXT_VERSION="2.14.6" ; yarn global add nuxt@$NUXT_VERSION` for example.

In both production and development scenarios the following env variables are respected:

```bash
NODE_ENV=production
NUXT_HOST=0.0.0.0
NUXT_PORT=3000
```

> The other David's comment about using Nginx in the compose it's true that is not needed, but we did it for 2 reasons, first to learn how to do it and second it obligated us to create another Dockerfile that will be the one used in k8s. Then I think it's not a bad practice to have this docker-compose to push the image to the registry, for example. For development purposes you have to use the `frontoffice` service and skip Nginx. But we will create pipelines that will push the Nginx image to the registry and we will use it in the k8s.

Then I think that a nice thing to do could be combining both the frontoffice and backoffice into this repository (that we can rename to anem-per-feina-frontend).

This will be a more interesting thing to do with Nginx, you can create a virtual host for the `/admin/**` route to redirect all the requests hitting that endpoint to a certain service where the backoffice will be running, and the rest of the routes will be handled by the frontoffice. This way it will make sense to use Nginx both for development and production.

**Off-topic question:**

Regardless of the learning exercise, would it be necessary to use Nginx in production for a single service, or it can be done by configuring the k8s Ingress with the desired parameters? In the end, the ingress can be an Nginx, right?

**Comment by jbagot** [comment]
Ok, thank you very much to clarify it. We will do it. 
Ingress can be nginx yes. I think we use traefik but can be nginx as well.

**Comment by jbagot** [comment]
The docker-compose for dev is finished. If you do a change in the code automatically is changed because we are using yarn dev.
We are using only docker-compose without Dockerfile.
Now we have to do production with a Dockerfile and using nginx with frontoffice and backoffice.
We changed the structure of the project a little bit

**Comment by jbagot** [comment]
I did the rebase to use the same structure as it is in master.
1. The docker-compose is for dev frontoffice. It has a volume and is using nuxt dev. Then if you change something in the code automatically is reloaded.
2. I created a production docker-compose with the frontoffice and nginx.

What can we do later? integrate the backoffice. I'm not sure if the backoffice is ready to be dockerized. But you can follow the same as we did in the frontoffice.

**Review by eudago — APPROVED** [review]

**Comment by d-asensio** [comment]
LGTM!

**Review by d-asensio — APPROVED** [review]

**Pull request merged**